streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -623,16 +613,46 @@ void runOnce() {
             return;
         }
 
-        initializeAndRestorePhase();
+        // we need to first add closed tasks and then created tasks to work with those revived / recycled tasks
+        restoreThread.addClosedTasks(taskManager.drainRemovedTasks());
+
+        // try to initialize created tasks that are either newly assigned or re-created from corrupted tasks
+        final List<AbstractTask> initializedTasks;
+        if (!(initializedTasks = taskManager.tryInitializeNewTasks()).isEmpty()) {
+            if (log.isDebugEnabled()) {
+                log.debug("Initializing newly created tasks {} under state {}",
+                        initializedTasks.stream().map(AbstractTask::id).collect(Collectors.toList()), state);
+            }
+
+            restoreThread.addInitializedTasks(initializedTasks);
+        }
+
+        // try complete restoration if there are any restoring tasks
+        if (taskManager.tryToCompleteRestoration(restoreThread.completedChangelogs())) {
+            log.debug("Completed restoring all tasks now");
+        }
+
+        if (state == State.PARTITIONS_ASSIGNED && taskManager.allTasksRunning()) {
+            // it is possible that we have no assigned tasks in which case we would still transit state
+            setState(State.RUNNING);
+
+            log.debug("All tasks are now running and transited State to {}", State.RUNNING);
+        }
 
-        // TODO: we should record the restore latency and its relative time spent ratio after
-        //       we figure out how to move this method out of the stream thread
-        advanceNowAndComputeLatency();
+        // check if restore thread has encountered TaskCorrupted exception; if yes
+        // rethrow it to trigger the handling logic
+        final TaskCorruptedException e = restoreThread.nextCorruptedException();
+        if (e != null) {
+            throw e;
+        }
 
         int totalProcessed = 0;
         long totalCommitLatency = 0L;
         long totalProcessLatency = 0L;
         long totalPunctuateLatency = 0L;
+
+        // TODO: we should allow active tasks processing even if we are not yet in RUNNING
+        //       after restoration is moved to the other thread
```

Comment:
We do have a JIRA ticket for this, and I do plan to do it right after this PR (since this is really a pre-requisite to have that). I will link the ticket here.